File: Algorithm/238. Product of Array Except Self.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-08-27 20:24:48
# @Last modified by:   WuLC
# @Last Modified time: 2016-08-27 20:26:14
# @Email: liangchaowu5@gmail.com

# O(n) time, O(n) space without division
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        left, right, result = [], [], []
        tmp, n = 1, len(nums)
        for i in xrange(n - 1):
            tmp *= nums[i]
            left.append(tmp)
        tmp = 1
        for i in reversed(xrange(1, n)):
            tmp *= nums[i]
            right.append(tmp)
        for i in xrange(n):
            if i==0:
                result.append(right[-1])
            elif i==n-1:
                result.append(left[-1])
            else:
                result.append(left[i-1]*right[n-i-2])
        return result
                
            

File: Algorithm/Java/238. Product of Array Except Self.java
/*
* @Author: WuLC
* @Date:   2017-06-23 09:36:40
* @Last Modified by:   WuLC
* @Last Modified time: 2017-06-23 09:42:41
* @Email: liangchaowu5@gmail.com
*/

// O(n) time, O(1)
// calculate left accumulated sum for each number firstly, then multiply by its right accumulated sum
public class Solution 
{
    public int[] productExceptSelf(int[] nums) 
    {
        int[] result = new int[nums.length];
        result[0] = 1;
        for(int i = 1; i < nums.length; i++)
            result[i] = result[i-1]*nums[i-1];
        int right = 1;
        for (int i = nums.length-1; i >= 0; i--)
        {
            result[i] *= right;
            right *= nums[i];
        }
        return result;
    }
}